### PR TITLE
Make libtest logfile respect --format

### LINKED ISSUE #57147: `--logfile` for libtest doesn't respect `--format`.
**Observed behaviour**
When using for example `cargo bench -- -Z unstable-options --format json --logfile bench.json`, `bench.json` has the usual human output format. Stdout prints JSON correctly though.

**Expected behaviour**
I expect to see JSON output in the file, exactly like in stdout.

**Notes**
I know at the moment rust is working on the custom test framework (https://github.com/rust-lang/rust/issues/50297), but other people are still relying on the current features (https://github.com/rust-lang/rust/issues/51924, https://github.com/rust-lang/cargo/issues/5670).
In any case, I think this is a bug, right? Would love to make a PR.

**Rustc**
rustc 1.33.0-nightly (a7be40c65 2018-12-26)

### PR DESCRIPTION
**Pull request by felipeamp**
I recently ran into an issue caused by the libtest output in logfile not respecting the --format flag. I've seen that a bug has been filed (#57147) but no PR was ever sent.

This PR makes the logfile follow the same format used for STDOUT printing. It removes the log_out field from ConsoleTestState and inlines its creation where needed, wrapping the logfile in an OutputFormatter. I've also deleted a couple of (now unused) methods in ConsoleTestState.

Let me know if you think there is a cleaner solution than the direction I've followed. Thanks!

closes #57147

### CODE REVIEW

**Comment by felipeamp** [comment]
Hi Jane,

Did I forget to do something to mark this PR as ready for review? Since this is my first PR to the Rust codebase I don't know if I missed some steps. Let me know if that's the case :)

Thanks!

**Comment by yaahc** [comment]
> Hi Jane,
> 
> Did I forget to do something to mark this PR as ready for review? Since this is my first PR to the Rust codebase I don't know if I missed some steps. Let me know if that's the case :)
> 
> Thanks!

No, sorry, you didn't do anything wrong. I've just behind on my reviews and actually planning on taking a break from the review rotation. Sorry for the delay, I'll reassign the PR.

r? rust-lang/libs

**Comment by thomcc** [comment]
I think this counts as a behavior change, and thus falls under t-libs-api rather than t-libs. So I'm going to reassign. This may be an overabundance of caution, though.

r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Comment by felipeamp** [comment]
Hey, is the process for a T-libs-api contribution the same as a T-libs one or do I need to do something else before this PR can be reviewed? :)

Thanks!

**Comment by JohnCSimon** [comment]
triage: 

@m-ou-se - can you answer this?

**Comment by m-ou-se** [comment]
Changes to the flags of `test` are a public API change, so require an FCP. Can you summarize the publicly visible change, maybe with a clear "before vs after" showcase?

Can we make this change without breaking any users?

**Comment by felipeamp** [comment]
Hi Mara,

Thanks for the reply!

The behavior only changes when a user runs a test or benchmark and sets both the `--format` and the `--logfile` flags. Currently the content printed to logfile is just a list of test case names with the strings "ok" or "failed" printed before them (depending on each test case result); while the content printed to STDOUT is always in the format given by the `--format` flag. This PR makes the content printed to the logfile also be in the format specified in the --format flag, just like STDOUT.

### Example:

Create a new Rust project with a single file `./src/lib.rs` with the following content:
```
#[test]
fn it_works() {
  let result = 2 + 2;
  assert_eq!(result, 4);
}

#[test]
fn it_does_not_work() {
  let result = 2 + 2;
  assert_eq!(result, 5);
}
```

Now run the following command:
`cargo +nightly  test -- -Z unstable-options --format=json --logfile=logfile.txt 1>stdout.txt`


**Current behavior:**

`logfile.txt` contains:
```
ok it_works
failed it_does_not_work
```

`stdout.txt` contains:
```
{ "type": "suite", "event": "started", "test_count": 2 }
{ "type": "test", "event": "started", "name": "it_does_not_work" }
{ "type": "test", "event": "started", "name": "it_works" }
{ "type": "test", "name": "it_works", "event": "ok" }
{ "type": "test", "name": "it_does_not_work", "event": "failed", "stdout": "thread 'it_does_not_work' panicked at 'assertion failed: `(left == right)`\n  left: `4`,\n right: `5`', src/lib.rs:10:3\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n" }
{ "type": "suite", "event": "failed", "passed": 1, "failed": 1, "ignored": 0, "measured": 0, "filtered_out": 0, "exec_time": 0.000431658 }
```


**New behavior:**
Both `stdout.txt` and `logfile.txt` contain:
```
{ "type": "suite", "event": "started", "test_count": 2 }
{ "type": "test", "event": "started", "name": "it_does_not_work" }
{ "type": "test", "event": "started", "name": "it_works" }
{ "type": "test", "name": "it_works", "event": "ok" }
{ "type": "test", "name": "it_does_not_work", "event": "failed", "stdout": "thread 'it_does_not_work' panicked at 'assertion failed: `(left == right)`\n  left: `4`,\n right: `5`', src/lib.rs:10:3\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n" }
{ "type": "suite", "event": "failed", "passed": 1, "failed": 1, "ignored": 0, "measured": 0, "filtered_out": 0, "exec_time": 0.000431658 }
```

 ### Re: Can we make this change without breaking any users?

Currently this change breaks users who rely on the current logfile output format.

I've thought about protecting this change behind a flag, thus not breaking users who rely on the current logfile behavior. The problem is that in this PR I've restructured the code to reuse the stdout-writing logic for the logfile writing as well, but the old logfile output didn't follow any specific format (as per the `--format` flag). Therefore I don't know if it's possible to restructure the code in a way that both the old and new behavior co-exist (at least I don't easily see how to do it). I could spend a few hours thinking hard about this if you'd like me to try. Just let me know how essential you think this is.



Please let me know how to proceed from here, especially since I don't know how the FCP process works :)

Thanks!

**Comment by JohnCSimon** [comment]
@m-ou-se - triage: can you help with this?
> Please let me know how to proceed from here, especially since I don't know how the FCP process works :)

**Comment by Manishearth** [comment]
cc @calebcartwright since you were looking in to doing more libtest stuff and may be able to review this

**Comment by hlopko** [comment]
Friendly ping @m-ou-se or @calebcartwright, it would be great to get this merged, thank you!

**Comment by Manishearth** [comment]
cc @rust-lang/testing-devex

**Comment by Manishearth** [comment]
@epage Hm. I think that's fair, but this is effectively fixing a bug in existing functionality, and the requisite work to make it possible to shift focus over hasn't happened yet. This PR has been open for more than a year, I really don't think it should wait on all that happening.

**Comment by epage** [comment]
If people have determined this to be a bug and want to move forward with merging it, go for it.  As for testing-devex, we've not done any prerequisite work to make any determinations like that on our side.

**Comment by jieyouxu** [comment]
cc @felipeamp from triage: I am marking this PR as waiting-on-author because merge conflicts.
@rustbot author

**Comment by Dylan-DPC** [comment]
@felipeamp any updates on this? thanks

**Comment by felipeamp** [comment]
I'm a bit busy in the next couple of weeks (going on vacation and all), but will fix the merge conflicts around Easter.

**Comment by JohnCSimon** [comment]
@felipeamp triage - pinging again, to be sure you haven't forgotten this 🙂

**Comment by felipeamp** [comment]
Hi John,

I've taken a look and the merge conflict (and moving this PR forward in general) requires a bit more work than I have the capacity at the moment. Luckily [aspotashev](https://github.com/aspotashev) has taken over the work in https://github.com/rust-lang/rust/pull/123365, which is current under review by `T-testing-devex` folks.

Thanks

**Comment by Dylan-DPC** [comment]
Closing this as it is superseded by #123365
